pytensor/tensor/einsum.py

```diff
@@ -417,6 +417,17 @@ def _right_to_left_path(n: int) -> tuple[tuple[int, int], ...]:
     return tuple(pairwise(reversed(range(n))))
 
 
+def _ensure_not_equal(elements):
+    """
+    Ensures that any pair in a list of elements are not the same object. If a pair of elements is found to be equal, then one of them is converted to a copy.
+    """
+    for i in range(len(elements)):
+        for j in range(i + 1, len(elements)):
+            if elements[i] == elements[j]:
+                elements[j] = elements[j].copy()
+    return elements
```

Comment:
Some small optimization. Don't assume a list is passed and reduce list access. Also `is` should be better than `==`.

```suggestion
    elements = list(elements)
    n_elem = len(elements)
    for i, elem1 in enumerate(elements[-1])
        for j, elem2 in enumerate(elements[i + 1:], start=i+1):
            if elem1 is elem2:
                elements[j] = elem1.copy()
    return elements
```